src/maker/api.rs

```diff
@@ -674,10 +717,32 @@ pub(crate) fn check_for_idle_states(maker: Arc<Maker>) -> Result<(), MakerError>
 
 /// Broadcast Incoming and Outgoing Contract transactions & timelock contract after maturity or hashlock contract.
 /// Remove contract transactions from the wallet.
-pub(crate) fn recover_from_swap(maker: Arc<Maker>) -> Result<(), MakerError> {
+pub(crate) fn recover_from_swap(
+    maker: Arc<Maker>,
+    mut tracker_stream: Option<Arc<TcpStream>>,
+) -> Result<(), MakerError> {
     // Broadcast all the incoming contracts.
     let (incomings, outgoings) = maker.wallet.read()?.find_unfinished_swapcoins();
     let is_hash_preimage_known = incomings.iter().any(|ic_sc| ic_sc.is_hash_preimage_known());
+    let timelock = outgoings
+        .first()
+        .map(|og_sc| og_sc.get_timelock().unwrap())
+        .unwrap();
+
+    // Block wait time is varied between prod. and test builds.
+    let block_wait_time = if cfg!(feature = "integration-test") {
+        Duration::from_secs(10)
+    } else {
+        Duration::from_secs(10 * 60)
+    };
+
+    let time_till_timelock_expires =
+        Duration::from_secs(timelock as u64 * block_wait_time.as_secs());
+    let start_time = Instant::now();
+    while !is_hash_preimage_known && start_time.elapsed() < time_till_timelock_expires {
+        check_for_tracker_watch_response(&maker, tracker_stream.as_mut().unwrap())?;
+        sleep(Duration::from_secs(10));
+    }
 
     if is_hash_preimage_known {
```

Comment:
We should need to refresh this flag here? As the flag is being set inside `check_for_tracker_watch_response` so once the response got back, there will be hashlocks thats are known.